src/compiler/performance.ts

```diff
@@ -0,0 +1,107 @@
+         * to generate heap snapshots.
+         *
+         * @param eventName A name for the event.
+         */
+        export function emit(eventName: string) {
+            if (profilerEvent) {
+                profilerEvent(eventName);
+            }
+        }
+
+        /**
+         * Increments a counter with the specified name.
+         *
+         * @param counterName The name of the counter.
+         */
+        export function increment(counterName: string) {
+            if (counters) {
+                counters[counterName] = (getProperty(counters, counterName) || 0) + 1;
+            }
+        }
+
+        /**
+         * Gets the value of the counter with the specified name.
+         *
+         * @param counterName The name of the counter.
+         */
+        export function getCount(counterName: string) {
+            return counters && getProperty(counters, counterName) || 0;
+        }
+
+        /**
+         * Marks the start of a performance measurement.
+         */
+        export function mark() {
+            return measures ? markInternal() : 0;
+        }
+
+        /**
+         * Adds a performance measurement with the specified name.
+         *
+         * @param measureName The name of the performance measurement.
+         * @param marker The timestamp of the starting mark.
+         */
+        export function measure(measureName: string, marker: number) {
+            if (measures) {
+                measures[measureName] = (getProperty(measures, measureName) || 0) + (Date.now() - marker);
+            }
+        }
+
+        /**
+         * Iterate over each measure, performing some action
+         * 
+         * @param cb The action to perform for each measure
+         */
+        export function forEachMeasure(cb: (measureName: string, duration: number) => void) {
+            return forEachKey(measures, key => cb(key, measures[key]));
+        }
+
+        /**
+         * Gets the total duration of all measurements with the supplied name.
+         *
+         * @param measureName The name of the measure whose durations should be accumulated.
+         */
+        export function getDuration(measureName: string) {
+            return measures && getProperty(measures, measureName) || 0;
+        }
+
+        /** Enables (and resets) performance measurements for the compiler. */
+        export function enable() {
+            counters = { };
+            measures = {
+                "I/O Read": 0,
+                "I/O Write": 0,
+                "Program": 0,
+                "Parse": 0,
+                "Bind": 0,
+                "Check": 0,
+                "Emit": 0,
+            };
+
+            profilerEvent = typeof onProfilerEvent === "function" && onProfilerEvent.profiler === true
+                ? onProfilerEvent
+                : undefined;
+            markInternal = performance && performance.now ? performance.now : Date.now ? Date.now : () => new Date().getTime();

```

Comment:
1. See comment below on `new Date().getTime()` performance (it should be `+new Date()` instead here).

2. Also please use `typeof performance!=='undefined'` instead of just checking for truthiness. Otherwise you'll get ReferenceError on platforms that don't have it defined.

3. Lastly, it's quite non-obvious whether **performance** identifier here resolves to global-scope performance or ts.performance. I can see you've got that `declare const performance` up at line 6 — but it feels like a trick, and it may come back biting as resolution rules change all the time.

  Perhaps a clearer solution would be to move the whole `Date.now` thing in together with other fallback-safe utils (forEach, contains, indexOf) in [core.ts around line 84](https://github.com/Microsoft/TypeScript/blob/master/src/compiler/core.ts#L79)?

  ```typescript
  namespace ts {
      export var preciseTime =
          typeof performance!=='undefined' && performance && performance.now ? performance.now :
          Date.now ? Date.now :
          () => +new Date();
  }
  ```

  That way rather than having defined and called markInternal here in performance.ts, you'd call `ts.preciseTime()` from core.ts. Also you would replace overly optimistic Date.now at [shims.ts isCancellationRequested at line 426](https://github.com/Microsoft/TypeScript/blob/master/src/services/shims.ts#L425) and [editorServices.ts resolveNamesWithLocalCache line 145](https://github.com/Microsoft/TypeScript/blob/master/src/server/editorServices.ts#L120)